File: C#/CosmosDB-NoSQL_CognitiveSearch/Service/CognitiveSearchService.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Azure;
using Azure.AI.OpenAI;
using Azure.Search.Documents;
using Azure.Search.Documents.Indexes;
using Azure.Search.Documents.Indexes.Models;
using Azure.Search.Documents.Models;
using Microsoft.Extensions.Configuration;

namespace CosmosRecipeGuide
{
    internal class CognitiveSearchService
    {
        private const string _SemanticSearchConfigName = "my-semantic-config";

        private readonly SearchIndexClient? _indexClient=null;
        private readonly SearchClient? _searchClient = null;

        private readonly string _searchIndexName =string.Empty;
        private readonly IConfiguration? _configuration =null;

        public CognitiveSearchService(IConfiguration config)
        {
            _configuration = config;
            
            _searchIndexName = _configuration["SearchIndexName"];

            _indexClient = CreateSearchIndexClient();
            _searchClient = _indexClient.GetSearchClient(_searchIndexName);
        }

        //check if the  index already exists
        public bool CheckIndexIfExists()
        {
            try
            {
                if (_indexClient.GetIndex(_searchIndexName) != null)
                {
                    return true;
                }
            }
            catch (RequestFailedException e) when (e.Status == 404)
            {
                return false;
            }

            return false;
        }

        //build the index
        public void BuildIndex()
        {
            DeleteIndexIfExists(_searchIndexName, _indexClient);
            CreateIndex(_searchIndexName, _indexClient);
        }

        // Upload documents as a batch.
        public async void UploadandIndexDocuments(List<Recipe> Recipes)
        {
            await _searchClient.IndexDocumentsAsync(IndexDocumentsBatch.Upload(ConvertRecipeToCogSarchDoc(Recipes)));
            
        }

        //Reducing the number of attributes in the class
        private List<CogSearchDoc> ConvertRecipeToCogSarchDoc(List<Recipe> recipes)
        {
            List<CogSearchDoc> cogSearchdocs=new List<CogSearchDoc>();
            foreach (Recipe recipe in recipes)
            {
                CogSearchDoc cdoc=new CogSearchDoc();
                cdoc.id = recipe.id;
                cdoc.name = recipe.name;
                cdoc.description = recipe.description;  
                cdoc.embedding  = recipe.embedding;

                cogSearchdocs.Add(cdoc);
            }

            return cogSearchdocs;
        }


        internal async Task<List<string>> SingleVectorSearch(float[] queryEmbeddings, int searchCount = 5)
        {

            // Perform the vector similarity search  
            var vector = new SearchQueryVector { K = 3, Fields = "embedding", Value = queryEmbeddings.ToArray() };
            var searchOptions = new SearchOptions
            {
                Vector = vector,
                Size = searchCount,
                Select = { "id"},
            };

            SearchResults<SearchDocument> response = await _searchClient.SearchAsync<SearchDocument>(null, searchOptions);

            List<string> ids = new List<string>();
            await foreach (SearchResult<SearchDocument> result in response.GetResultsAsync())
            {
                ids.Add(result.Document["id"].ToString());    
            }
            return ids; 
        }



        private SearchIndexClient CreateSearchIndexClient()
        {
            string searchServiceEndPoint = _configuration["SearchServiceEndPoint"];
            string adminApiKey = _configuration["SearchServiceAdminApiKey"];

            SearchIndexClient _indexClient = new SearchIndexClient(new Uri(searchServiceEndPoint), new AzureKeyCredential(adminApiKey));
            return _indexClient;
        }

        private SearchClient CreateSearchClientForQueries(string indexName)
        {
            string searchServiceEndPoint = _configuration["SearchServiceEndPoint"];
            string queryApiKey = _configuration["SearchServiceQueryApiKey"];

            SearchClient _searchClient = new SearchClient(new Uri(searchServiceEndPoint), indexName, new AzureKeyCredential(queryApiKey));
            return _searchClient;
        }

              
        private void DeleteIndexIfExists(string indexName, SearchIndexClient _indexClient)
        {
            try
            {
                if (_indexClient.GetIndex(indexName) != null)
                {
                    _indexClient.DeleteIndex(indexName);
                }
            }
            catch (RequestFailedException e) when (e.Status == 404)
            {
                return;
            }
        }

        private void CreateIndex(string indexName, SearchIndexClient _indexClient)
        {
            _indexClient.CreateOrUpdateIndex(BuildVectorSearchIndex(indexName));
        }

        internal SearchIndex BuildVectorSearchIndex(string name)
        {
            string vectorSearchConfigName = "my-vector-config";

            SearchIndex searchIndex = new(name)
            {
                VectorSearch = new()
                {
                    AlgorithmConfigurations =
                {
                    new VectorSearchAlgorithmConfiguration(vectorSearchConfigName, "hnsw")
                }
                },
                SemanticSettings = new()
                {

                    Configurations =
                    {
                       new SemanticConfiguration(_SemanticSearchConfigName, new()
                       {
                           TitleField = new(){ FieldName = "name" },
                           ContentFields =
                           {
                               new() { FieldName = "description" }
                           }
                       })

                },
                },
                Fields =
            {
                new SimpleField("id", SearchFieldDataType.String) { IsKey = true, IsFilterable = true, IsSortable = true, IsFacetable = true },
                new SearchableField("name") { IsFilterable = true, IsSortable = true },
                new SearchableField("description") { IsFilterable = true },
                new SearchField("embedding", SearchFieldDataType.Collection(SearchFieldDataType.Single))
                {
                    IsSearchable = true,
                    Dimensions = 1536,
                    VectorSearchConfiguration = vectorSearchConfigName
                }
            }
            };
            return searchIndex;
        }
         
    }
}

File: Java/CosmosDB-NoSQL_CognitiveSearch/Code/src/main/java/com/azure/recipe/service/CognitiveSearchService.java
package com.azure.recipe.service;

import com.azure.core.credential.AzureKeyCredential;
import com.azure.core.util.Context;
import com.azure.recipe.AppConfig;
import com.azure.recipe.model.CogSearchDoc;
import com.azure.recipe.model.Recipe;
import com.azure.search.documents.SearchClient;
import com.azure.search.documents.indexes.SearchIndexClient;
import com.azure.search.documents.indexes.SearchIndexClientBuilder;
import com.azure.search.documents.indexes.models.*;
import com.azure.search.documents.models.SearchOptions;
import com.azure.search.documents.models.SearchQueryVector;
import com.azure.search.documents.util.SearchPagedIterable;
import lombok.extern.slf4j.Slf4j;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Slf4j
public class CognitiveSearchService {
    private static final String MY_SEMANTIC_CONFIG = "my-semantic-config";
    private String searchIndexName;
    private SearchIndexClient indexClient = null;
    private SearchClient searchClient = null;


    public CognitiveSearchService() {
        searchIndexName = AppConfig.searchIndexName;
        indexClient = createSearchIndexClient();
        searchClient = indexClient.getSearchClient(searchIndexName);
    }

    private SearchIndexClient createSearchIndexClient() {
        String searchServiceEndPoint = AppConfig.searchServiceEndPoint;
        String adminApiKey = AppConfig.searchServiceAdminApiKey;

        return new SearchIndexClientBuilder()
                .endpoint(searchServiceEndPoint)
                .credential(new AzureKeyCredential(adminApiKey))
                .buildClient();
    }

    public boolean checkIndexIfExists() {
        try {
            if (indexClient.getIndex(searchIndexName) != null) {
                return true;
            }
        } catch (Exception e) {
            log.error("Index does not exist", e);
            return false;
        }
        return false;
    }

    public List<String> singleVectorSearch(List<Float> queryEmbeddings) {

        var vector = new SearchQueryVector();
        vector.setKNearestNeighborsCount(3);
        vector.setFields("embedding");
        vector.setValue(queryEmbeddings);

        var searchOptions = new SearchOptions();
        searchOptions.setVector(vector);
        searchOptions.setSelect("id");
        searchOptions.setTop(5);

        SearchPagedIterable response = searchClient.search(null, searchOptions, Context.NONE);

        return response
                .stream()
                .map(result -> (String) result.getDocument(Map.class).get("id"))
                .collect(Collectors.toList());
    }

    //build the index
    public void buildIndex() {
        deleteIndexIfExists(searchIndexName);
        createIndex(searchIndexName);
    }

    void uploadAndIndexDocuments(List<Recipe> Recipes) {
        List<CogSearchDoc> documents = convertRecipeToCogSarchDoc(Recipes);
        searchClient.indexDocuments(new IndexDocumentsBatch().addUploadActions(documents));

    }

    public void uploadandIndexDocuments(List<Recipe> Recipes) {
        IndexDocumentsBatch batch = new IndexDocumentsBatch()
                .addUploadActions(convertRecipeToCogSarchDoc(Recipes));

        searchClient.indexDocuments(batch);
    }

    private List<CogSearchDoc> convertRecipeToCogSarchDoc(List<Recipe> recipes) {
        return recipes
                .stream()
                .map(recipe -> {
                    CogSearchDoc cdoc = new CogSearchDoc();
                    cdoc.setId(recipe.getId());
                    cdoc.setName(recipe.name);
                    cdoc.setDescription(recipe.description);
                    cdoc.setEmbedding(recipe.embedding);
                    return cdoc;
                }).collect(Collectors.toList());

    }

    private void deleteIndexIfExists(String indexName) {
        if (indexClient.getIndex(indexName) != null) {
            indexClient.deleteIndex(indexName);
        }
    }

    private void createIndex(String indexName) {
        indexClient.createOrUpdateIndex(buildVectorSearchIndex(indexName));
    }

    private SearchIndex buildVectorSearchIndex(String name) {
        String vectorSearchConfigName = "my-vector-config";

        SearchIndex searchIndex = new SearchIndex(name);
        VectorSearch vectorSearch = new VectorSearch();
        vectorSearch.setAlgorithmConfigurations(List.of(new HnswVectorSearchAlgorithmConfiguration(vectorSearchConfigName)));

        searchIndex.setVectorSearch(vectorSearch);
        SemanticSettings semanticSettings = new SemanticSettings();
        PrioritizedFields prioritizedFields = new PrioritizedFields();

        SemanticField titleField = new SemanticField();
        titleField.setFieldName("name");

        prioritizedFields.setTitleField(titleField);
        SemanticField contentField = new SemanticField();
        contentField.setFieldName("description");
        prioritizedFields.setPrioritizedContentFields(List.of(contentField));

        semanticSettings.setConfigurations(List.of(new SemanticConfiguration(MY_SEMANTIC_CONFIG, prioritizedFields)));
        searchIndex.setSemanticSettings(semanticSettings);

        SearchField idSearchField = new SearchField("id", SearchFieldDataType.STRING);
        idSearchField.setKey(true);
        idSearchField.setFilterable(true);
        idSearchField.setSortable(true);

        SearchField nameSearchField = new SearchField("name", SearchFieldDataType.STRING);
        nameSearchField.setFilterable(true);
        nameSearchField.setSortable(true);
        nameSearchField.setSearchable(true);

        SearchField descSearchField = new SearchField("description", SearchFieldDataType.STRING);
        descSearchField.setFilterable(true);
        descSearchField.setSearchable(true);

        SearchField embedingSearchField = new SearchField("embedding", SearchFieldDataType.collection(SearchFieldDataType.SINGLE));
        embedingSearchField.setSearchable(true);
        embedingSearchField.setVectorSearchDimensions(1536);
        embedingSearchField.setVectorSearchConfiguration(vectorSearchConfigName);
        searchIndex.setFields(List.of(
                idSearchField,
                nameSearchField,
                descSearchField,
                embedingSearchField
        ));

        return searchIndex;
    }


}
